### ISSUE TITLE
Newly added door entities - collision problem

### ISSUE THREAD

**Issue by Const1702** [MattFiler/OpenCAGE#95]
When a composite entity is added through the Cathode editor with the path **AYZ\DOORS\<Any door type>** the door does show and it has its functions, but the collisions are a problem:

1. It doesn't let the player pass through, even when it is open.
2. Bullets also don't go through and flares bounce off the opened part of the door.

There are **no problems** with the already **existing doors** on the level, only with the added ones.

Attempted on level: **Tech_Hub**
The door composites are added to AYZ\LEVELS\SEVASTOPOL\TECHNICAL\TRANSITSTATION\TRANSITSTATION_SECTION

The way I entered this level isn't by loading a mission, but I used the "Select Level" from the Debug Checkpoint mod, that way it completely refreshes and the changes from the editor can properly apply, but it still doesn't fix the collision problem with the new doors.

(I am not sure if this is an issue or it requires extra modifications to fix)

**OpenCAGE version: 0.9.3.02**

**Comment by MattFiler** [comment]
Unfortunately this is a known issue. I'm not entirely sure on the fix just yet, and think it's something I'm yet to reverse engineer. 

In the meantime, you may be able to get around it by modifying the CollisionBarrier entities within the door composite?

**Comment by Const1702** [comment]
Thank you for reading! I'll definitely try to play around with the CollisionBarrier. Hope it gets resolved soon!

**Comment by MattFiler** [comment]
Let me know if you have any success. The only other lead I have atm is that doors are based on a `Door` entity function type, which could be doing collision-related stuff internally. In which case, it would probably be that something collision-related is going through the Mover file, which I don't add entries for new doors within (so would explain the issue).

It's something that has been on the roadmap for a while and will be a focus soon!

**Comment by Const1702** [comment]
I was looking through the properties of a door composite (DOOR_MED to be exact), but I couldn't find anything related to collisions. Only when I click on "Door_Package_1" and then selecting "Go To". Then I can find 2 collision-related entities:
**- PhysicsBarrier (CollisionBarrier)**
**- ClosingBarrier (CollisionBarrier)**
I don't think these are useful, because they are related to door packages that are defining whenever the door has a lever, button, hacking, keycard, passcode or just a blank panel and that is a small portion of the door. I explored the "door" function entity and most of its features are triggers (Example: started_opening, finished_closing...), but the ones that grabbed my eye are: "geometry", "nav_mesh", "invert_nav_mesh_barrier", but they are still difficult for me to understand.

I have also found some small differences between existing and new doors. There are 2 that I know of:

1. New doors don't have a delayed animation. If you don't know what a delayed animation is, it is when both of the status lights (the red and green ones) turn on for 1 second, then the door makes a bloop sound and it opens. This animation is random when opening existing doors, but never appears when opening new doors. If you still don't understand I can post a video showing how the delayed animation looks like.
2. When door packages (Levers, buttons...) are changed through the editor for an existing doors, as another user in a closed issue mentioned, it does update the mechanism, but it also leaves the previous panel's texture floating in the air. This is an issue that you are aware of. In added doors the door package issue doesn't happen. It doesn't leave a floating panel and it changes the mechanism properly.

With those two issues shown, I think that there might be very small differences between the existing and added doors' mechanisms and animations. If you are going to put doors on your priority list, I recommend you check in a detailed way packages and animations.

**Comment by MattFiler** [comment]
The "delayed animation" you mention is actually an indicator that the game is loading in content behind the door. This loading is done via the scripting system's use of `Zone` entities, which use `TriggerSequence` entities to group objects to load in/out. When you approach a door, it can optionally wait for the zone contents to be loaded before opening, which allows for level streaming and optimisations to memory overhead at runtime. If you've just added a new door, you probably haven't hooked any zones up to it, which is why you won't get the delay as nothing is loading.

Door packages lingering are again related to the Mover descriptors - changing the package will break the link to the package listed in the MVR file, which will cause it to take all info from the MVR and become a static renderable in the world. You should be able to fix this in the new update by using the "Edit Movers" button and positioning the mover for the package far off in the level. It's not an ideal solution, but one that works for now while I better figure out the contents of the Mover descriptor so that I can add/remove them.

**Comment by MattFiler** [comment]
Made a discovery related to this today which I'll post here to keep my thoughts in one place:
- Swapping all internal resources to `resource` parameters produces the same collision issue with doors where you can't walk through them after opening.
- Is the issue with new doors perhaps an ID issue when pointing to a resource reference?

**Comment by Const1702** [comment]
I have a feeling that this may be related to NavMeshBarriers located in a door's package (in my case the door is AYZ\DOORS\DOOR_MED, I just searched door_package and click "Go To"). NavMeshBarriers has paramaters and resource references like shown in the pictures below: 
![image](https://user-images.githubusercontent.com/116276106/216350614-264055b5-b1bb-4086-814e-dbd2b15483af.png)
![image](https://user-images.githubusercontent.com/116276106/216350721-2aaaa989-ea94-4458-aeac-9f0ee5a1b5f4.png)
"COLLISION_MAPPING" and "NAV_MESH_BARRIER_RESOURCE" don't seem to be investigated enough to modify them.

Another thing I discovered is that again in the door_package there are 2 collisionbarriers:
![image](https://user-images.githubusercontent.com/116276106/216351367-cb438287-e914-45ae-8b04-567bbaff744f.png)
Both of these barriers have the resource reference "COLLISION_MAPPING". The ClosingBarrier seems to have a "Door" function entity connecting IN to it:
![image](https://user-images.githubusercontent.com/116276106/216351991-e5ac8322-db85-4e29-a13e-6d0b6639f0c7.png)

I have tried moving existing doors around the map before and never had problems. Only when creating new ones. Door_Package seems to be a crucial part of the door, because DataType FLOATs like "request_open", "force_open", "request_close" all seem to be connecting to Door_Package. The ModelReferences in the actual door (AYZ\DOORS\DOOR_MED for me) seem to have links connecting IN to them from the Door_Package and not from the actual "request_open", "force_close"...
![image](https://user-images.githubusercontent.com/116276106/216354399-811c1462-cb02-4276-ae5f-e9e2e301c36f.png)

**Comment by Const1702** [comment]
Something related to the lingering door mechanism problem I discovered is that when I move an existing door to a new position and then change its mechanism it does change it, but it leaves the lingering part not to its new position, but at the old one. Here is an example of moving the Door_SML leading to the transit control lift in the SCI_Hub level. What is happening here is that the broken mechanism is always rendered to the position that was initially assigned to it by the game devs, no matter where the existing door is moved:
![image](https://user-images.githubusercontent.com/116276106/216792070-1d30a55b-fee0-4d99-8b69-3e409fe6e6fd.png)
How I achieved this bizarre discovery:
If you are testing this it is very important to **move the existing door first** and **then change the mechanism**, so you can see how the mover not only gets detached, but also gets moved to the original position where the door was before.

With newly added doors the floating problem **doesn't occur at all** for some reason. I'm not sure if this is related to entity creation for doors.

You are saying that the current way to fix this problem is to move the mechanism away from the map. I'm not sure if I am doing it correctly, but here is the way I am trying to achieve this: Open the resource reference editor directly on the door -> Add new renderable instance -> Change Model  -> This is where I run into a problem. The DOOR_PACKAGE composite is located in ARCHETYPES\... and in the model browser's sidebar there isn't an "ARCHETYPES" list like in the main window:
![image](https://user-images.githubusercontent.com/116276106/216792357-7b6dc0e8-2faf-4832-86ca-fe24b415fabd.png)
Tried looking through the available lists, but found nothing. This is probably the wrong way of doing it and it might not be possible yet. But if it is, could you please let me know how to get control of the mechanisms (Door_Mechanism_Front and Door_Mechanism_Back)? You also mentioned that it is related to mover descriptors, but the movers button for the door in greyed out, so I can't control it from there. It seems to be only available in ModelReferences, LightReferences, FogSpheres (I think) and the only way to get to a door's ModelReference is to click "Go To" and modify it from there, but that will change it globally for all doors like this.
![image](https://user-images.githubusercontent.com/116276106/216792437-476b6725-af77-40ef-ba3d-bf65875421c4.png)

I was trying to use Overrides to delete door mechanisms with the "deleted" paramater, but that also didn't work, due to the checksum problem with overrides.

**Comment by Const1702** [comment]
I made a discovery with collisions. When an existing door is being hit with a wrench it makes a sound (can be also a glass sound if you are hitting its glass part with doors that have glass on them), but with new doors it does nothing. Flares and collisions work properly with both doors - they bounce off when closed, but when new doors are open flares keep bouncing off the opened part. Shooting at a closed existing door leaves a bullet hole and shooting a new one doesn't (the bullets might be still getting stopped by the door tho, because if I shoot it at an angle where it would go through the door an hit the ceiling it doesn't. That means it is probably only swallowing the bullet without a trace). Flamethrower fire collides with existing doors and goes through newly added ones when they are closed, like they are not there.

In the video shown below (on level: ENG_TowPlatform) the door in its frame is the existing one to the airlock and the badly positioned one is the newly added one: _File size too big so unfortunately it had to be done with YT_ **https://youtu.be/uMTc1oqzMqU**

**Comment by Const1702** [comment]
**Question related to collisions:** What does the blank collision mean? It is not selectable in the dropdown menu. You can find it in some existing ModelReferences. In this case Level: **ENG_TowPlatform**, Composite: **AYZ\DOORS\DOOR_MED**, Entity: **Door_R (ModelReference)**:
![image](https://user-images.githubusercontent.com/116276106/236823287-8c2f5e72-381c-42bb-b334-142c5cd7f577.png)

Is it an unidentified collision by OpenCAGE so newly added entities are created without it? I've noticed if I add some corridor/interior elements like technical ceilings, walls, floors they have no collision at all. You just fall through them (only with newly added ones). On ENG_TowPlatform -> **AYZ\TECHNICAL\CORRIDOR_ELEMENTS\FLOOR** (haven't tested this composite in this level to see if it works): **Collision01_COL (ModelReference)** has an blank collision mapping, but the actual **Floor (ModelReference)**, there is mapping assigned to it.

I remember seeing a twitter post (that I can't find right now) you made about someone making a custom level under a day not so long ago and he showed a door and going through and he had no collision problem with it. Also the interior/corridor elements in that level from the twitter post had collisions and the person showing it didn't fall through.

_Congratulations on making it on the game news (PC Gamer...) by the way!_

**Comment by MattFiler** [comment]
Ok, I've found some more out about this!

Doors contain a `PhysicsSystem` entity, which defines a `DYNAMIC_PHYSICS_SYSTEM` resource reference within the door composite, pointing to a physics system index. This resource reference is seemingly further defined by the `PHYSICS.MAP` file to define the physics system for the door. 

It's a similar system to how the `EnvironmentModelReference` nodes within doors define a `ANIMATED_MODEL` resource reference (for the door open/close anims), which point to an environment animation index seemingly further defined by the `ENVIRONMENT_ANIMATION.DAT` file.

While looking into this I've also figured out that the `COLLISION_MAPPING` resource references appear to point to the `COLLISION.MAP` file, which defines a load of info about the collision mapping for each instance of an entity (similar to the `PHYSICS.MAP` contents).

All of these resource definitions appear to then also be referenced within `RESOURCES.BIN` which is like a global container/database for all the resources in the level.

I'm not parsing or using any of the above files just yet, which explains why more complex composites that rely on resources are sometimes working incorrectly when created from scratch, and indeed why the doors don't work properly when added via a new composite instance rather than just moving existing ones. Commands is apparently only one piece of the puzzle when it comes to the data read-in by the engine.

I've figured all of this out off of the back of the work done for #194, which was caused by Character entities requiring a separate file to define their instanced data in a similar way to this. Figuring out how that extra file linked to commands has unearthed this plethora of new stuff to reverse engineer - and also is very similar to the issue in #144, in that the MVR file contains instance specific material overrides for instances of entities (which I also now know how to link up to commands thanks to #194).

Unforuntately, I don't think it's gonna be as simple as just reverse engineering the files, as the way I display data in the script editor currently is pretty poor in regards to understanding the actual hierarchies of instanced nodes - for example:
- I create a composite with a `ModelReference` entity in it positioned at 0,0,0
- I create another composite, and instance that first composite at position 1,1,1
- I add another instance of that first composite to my second composite at position -1,-1,-1
- I set my second composite as the root composite that the level loads to
If I was now to load the level, I'd have two `ModelReference` entities within it, one at 1,1,1 and one at -1,-1,-1. Both are coming from the same composite & are the same entity, but they are two instances within the runtime game.

My script editor is great at showing the composites individually, but it's bad at letting you get an understanding of how those hierarchies will populate at runtime. For example, what if I wanted to colour one instance of the `ModelReference` blue and another green? I could use overrides - but then understanding that hierarchy, particularly if I begin adding in additional levels of composites there is really confusing. This sort of thing is entirely what MVR is for - adding instance specific information to instanced entities. It's also what a lot of these additional files seem to be for - defining instance specific information for resources.

To display it better in my editor I need to do a lot of reworks, to be able to jump down hierarchies and have a better conceptual understanding of where you are (as well as just editing the actual composite itself to apply changes across the board). It'd be better backed up by a 3D interface too. And here you see my issue - as I try to implement one thing, the list grows and grows 😂

The dream would be a full-on prefab system similar to something you'd find in Unity.

I might be able to get away with doing some simpler auto-calculations for this data in the meantime, but first I need to reverse engineer the file formats & then have a think about how I could automate all of it in the background. I might end up just displaying it similar to how the MVR & Character instance-specific stuff is done currently, while I work on the nicer instance display.

TL;DR: I think there's a whole load of additional files to reverse engineer, notably `RESOURCES.BIN` and `PHYSICS.MAP` for the doors to function. It'll make the commands editor considerably more robust at generating levels, but will be a considerable amount of effort to display nicely!

**Comment by MattFiler** [comment]
In the spirit of keeping this issue updated - #268 has been created to cover the ground work for implementing the above changes.

**Comment by MattFiler** [comment]
Pinning this issue to mark that it's currently top priority.

**Comment by MattFiler** [comment]
I've been back looking at this again, and I've definitely narrowed it down to the PhysicsSystem entity. Its DYNAMIC_PHYSICS_SYSTEM resource (which actually gets applied to the entire composite) is defined via the PHYSICS.MAP file, which itself is then linked to commands via a separate RESOURCES.BIN file.

If I clear out the PHYSICS.MAP file, the game doesn't crash, but also none of the doors collisions work, in exactly the same way as if you add a new door. So this is definitely the missing info the game is after.

The file contains a big matrix for each entry, and if you mess with the matrix all the physics in the level goes bizarre - I'm assuming because colliders are flying around the place. Basically, I need to figure out what this matrix is defining, and then we can make new ones.

As always with this reverse engineering, this is now also presenting other issues, as I need to figure out what all the other stuff in RESOURCES.BIN is pointing to in order to add new entries to it as well. It's seemingly connected to MODELS.MVR, which is the bane of my existence :D

**Comment by MattFiler** [comment]
I'm now correctly writing out to RESOURCES.BIN, COLLISION.MAP, and PHYSICS.MAP - and newly added doors are working!

I'll tidy up the code and push it to staging soon.

**Comment by FoxyFoxy67** [comment]
Since the last update (0.14.0.029) I now have the same issue with the doors I add, for any types and any level.

Steps to reproduce the issue:
- Load a level
- Place a door
- Try to pass through the door

OpenCAGE Version: 0.14.0.029
OpenCAGE Branch: "staging"
(Edit: I forgot to specify which branch of OpenCAGE I'm on)

**Comment by MattFiler** [comment]
To keep this updated, I'm now quite close to having a working solution for this! I've implemented all the logic required to calculate the physics objects in the level which will mean newly added ones can be included. Just need to fix up some bits and do some further testing and then will get it out on to staging when it's ready :)

**Comment by MattFiler** [comment]
Appreciate it has been years at this point, but in the spirit of keeping ticket progress updated - I now have this fully working.

Unfortunately, re-writing this data requires a full rebake of the level (which is also now working), which means not only physics/collisions, but also sound networks, lighting, navmeshes, and other complex data like cover positions and spotting jobs for NPCs.

I have a working solution for all of them, but currently the lighting bake and calculations for the NPCs are not where I want them to be, so I'll continue iterating on that before I push it out to staging as it could majorly impact gameplay.

Hope to have something out by the end of the year!

**Comment by MattFiler** [comment]
And also here's proof! This is the SOLACE level completely re-instanced through OpenCAGE, which means it has fully regenerated:

- Movers (including material data, lighting data, all instance specific driven by the scripting logic)
- Collisions
- Physics
- Sound networks (and other properties such as reverbs and ambience)
- Navmesh
- NPC cover positions, spotting jobs, assault positions
- Lighting (radiosity data)
- Occlusion and lighting BVHs
- Alphalight

And obviously all the existing stuff including script logic, models, textures, and shaders. So yeah - literally everything built through the tools!

<img width="1920" height="1080" alt="Image" src="https://github.com/user-attachments/assets/2187eb62-978f-40fd-8326-e9352125fa98" />

**Comment by MattFiler** [comment]
Apologies again for how long this has taken but it really has been a stack of 'everything needs to work together' for just one new feature addition!

**Comment by FoxyFoxy67** [comment]
Hey,
Don't worry, the work you are doing is amazing, take your time.

Now I know why the new sound nodes I added in the level did not work, they need to be instanced too. (or maybe I had the wrong soundbank loaded)

**Comment by MattFiler** [comment]
This is now live on staging! To build a level with instanced data, click 'Save and Build' from the 'File' menu!

Please let me know if you run into any problems :)

**Comment by Const1702** [comment]
I checked OpenCAGE today and as soon as I saw "Save and Build" I sprinted to GitHub to see what is new. I am in AWE!!!! This is incredible and I cannot even comprehend the amount of work that was put into this. Just thinking for example about how Movers are calculated where the tool checks one parameter let's say which then defines whether an entity is "deleted" which then is checked to see if a mover should be deleted in favour of another one is so complex to me - yet you managed to make these calculations also for collisions, physics, everything accessible from one button and that amazes me! This made my next weeks seeing all the possibilities now since I can poke around with the new stuff and perhaps report any issues I find or ask questions since I unfortunately missed lots of new features being implemented.

Sadly, mid-2025 until mid-2026 has been an extremely busy period of time for me but I hope for the best in terms of free time for this month and onward 🙂. Seeing `Fixed In Staging Branch` on this issue feels surreal and I love it! I have always admired the environment design and scripting structure this game has to offer.

What is a mystery to me is how all GitHub notification emails from my "GitHub" Gmail label stopped actually notifying me for a couple months now. I had to re-create that label earlier today and activity restarted once again which is good.

Currently, I have been messing around with doors (obviously) and noticed a little complication which seems to be occurring after I rebuild a level. That is the `BackstageAlien` (probably the `Xenomorph_NPC` too but not sure) going permanently "backstage" and refusing to come out of the vents while clearly getting aggravated by noise and sense. I guess it is a little shy 😄! Well, I will investigate this deeper today and write a specific issue if I see the problem. Might be navmesh-related but I see some new staging updates arrived and need to recheck it. I just am a bit hesitant to add to the workload right now since you obviously deserve a break after implementing all these algorithms!

Once again, apologies for the abrupt disappearance and my lack of reaction to the many comments you passionately added to this issue!
